lib/rules/no-invalid-position-declaration/index.mjs

```diff
@@ -0,0 +1,60 @@
+import findNodeUpToRoot from '../../utils/findNodeUpToRoot.mjs';
+import isStandardSyntaxDeclaration from '../../utils/isStandardSyntaxDeclaration.mjs';
+import { nestingSupportedAtKeywords } from '../../reference/atKeywords.mjs';
+import report from '../../utils/report.mjs';
+import ruleMessages from '../../utils/ruleMessages.mjs';
+import validateOptions from '../../utils/validateOptions.mjs';
+
+const ruleName = 'no-invalid-position-declaration';
+
+const messages = ruleMessages(ruleName, {
+	rejected: 'Unexpected invalid position declaration',
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/no-invalid-position-declaration',
+};
+
+/** @type {import('stylelint').CoreRules[ruleName]} */
+const rule = (primary) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, { actual: primary });
+
+		if (!validOptions) return;
+
+		root.walkDecls((decl) => {
+			if (!isStandardSyntaxDeclaration(decl)) return;
+
+			const { parent } = decl;
+
+			// Declarations are allowed in rules
+			if (parent?.type === 'rule') return;
+
+			if (parent?.type === 'atrule') {
+				const atRuleName = parent?.name.toLowerCase();
+
+				// Nesting-supported at-rules only allow declarations when nested inside a rule
```

Comment:
question

Are you sure?

> [`@apply --three {color: red;}`](https://drafts.csswg.org/css-mixins/#apply-rule)

Reply:
I interpret the spec as `@apply` only being valid inside rules:

> The [`@apply`](https://drafts.csswg.org/css-mixins/#at-ruledef-apply) rule is only valid in the body of a [style rule](https://drafts.csswg.org/css-syntax-3/#style-rule) or [nested group rule](https://drafts.csswg.org/css-nesting-1/#nested-group-rules); using it in any other context causes it to be invalid and ignored.

Reply:
Now I get it, that's what he meant by "when nested inside a rule"; I thought it was inside the _inner_ block.
i.e. checking the at-rule's parent node, is outside the scope of the rule if the declaration is already inside a rule
e.g.
```css
@apply --three {
  a {
    color: red;
  }
}
```